### PR TITLE
LLVM and SPIRV-LLVM-Translator pulldown (WW26 2024)

### PR DESCRIPTION
**Pull request by iclsrc**
LLVM: llvm/llvm-project@89d8df12015ac3440190d372a8d439614027dc2c
SPIRV-LLVM-Translator: KhronosGroup/SPIRV-LLVM-Translator@305f48884606abf

### CODE REVIEW

**Comment by frasercrmck** [comment]
https://github.com/llvm/llvm-project/commit/fa48dbdf3d6ca5f8f580671d572f141e788b1d9f is not a valid llvm-project commit.

**Comment by jsji** [comment]
> llvm/llvm-project@89d8df1
Thanks for reporting, yes, it should be  https://github.com/llvm/llvm-project/commit/89d8df12015ac3440190d372a8d439614027dc2c

**Comment by jsji** [comment]
@intel/dpcpp-cfe-reviewers Would you please have a look at follow testcase updates and template arg fix.

- [ ] [[SYCL] Update clang verification checks to use fully qualified class …](https://github.com/intel/llvm/pull/14327/commits/73b3c3f2e4734aa95dbd9d758abb08dde2298b0c)
- [ ] [[SYCL] Adjust additional test for clang verification](https://github.com/intel/llvm/pull/14327/commits/f48bae25ce1972790e2781d764a24d826db9cc46)
- [ ] [[SYCL][E2E] Fix Wmissing-template-arg-list-after-template-kw warnings](https://github.com/intel/llvm/pull/14327/commits/97e43dddcebfaa44214533888d072fa4cf3f24d5)
- [ ] [[SYCL][Test] Update test due to GEP folding change](https://github.com/intel/llvm/pull/14327/commits/a5f1d633013a35214aad9d47afadbb809ffa523b)
- [ ] [[Clang] Don't hardcode opnumber in module test](https://github.com/intel/llvm/pull/14327/commits/78207029992cdc6996e671c270e999cf1914aab9)

@intel/dpcpp-tools-reviewers Would you please have a look at the LLVM and tools update. Thanks.

- [ ] [[SYCL] Add PassInstrumentation header in ComputeModuleRuntimeInfo](https://github.com/intel/llvm/pull/14327/commits/209ca8eec603d0d8fe8cd041c72a7460007088de) 
- [ ] [[SYCL] Update sycl-offload-new-driver due to order diff](https://github.com/intel/llvm/pull/14327/commits/f2f3acc5c9441d88b3afdf2ccc902cde99f4ed0e)

@intel/dpcpp-spirv-reviewers Would you please have a look at the test update and XFAIL.

- [ ] [llvm-spirv: update test due to entry point diff](https://github.com/intel/llvm/pull/14327/commits/4e6a41426603373cd21c1a4102769b00e80b4022)
- [ ] [llvm-spirv: XFAIL RuntimeAligned.spt test](https://github.com/intel/llvm/pull/14327/commits/2515f6f5e1a3b15c3a9d5ef38421785d6a7541e3)

File: llvm-spirv/test/extensions/INTEL/SPV_INTEL_runtime_aligned/RuntimeAligned.spt
Comment (jsji) [inline llvm-spirv/test/extensions/INTEL/SPV_INTEL_runtime_aligned/RuntimeAligned.spt]: Also failing in sycl branch.

File: clang/test/Driver/sycl-offload-new-driver.c
Comment (jsji) [inline clang/test/Driver/sycl-offload-new-driver.c:9]: Somehow two `input` are swapped now.

File: clang/test/Modules/no-external-type-id.cppm
Comment (jsji) [inline clang/test/Modules/no-external-type-id.cppm:26]: The number is different due to sycl branch changes.

File: clang/test/Driver/sycl-offload-new-driver.c
Comment (sarnex) [reply clang/test/Driver/sycl-offload-new-driver.c:9]: fine for pulldown but i wonder are we supposed to honor the order the user provided in `-fsycl-targets=nvptx64-nvidia-cuda,spir64`? IMO either order should work exactly the same from the user POV so it shoudn't matter but @mdtoguchi any take? Thx

**Review by sarnex — APPROVED** [review]
lgtm for tools

File: llvm-spirv/lib/SPIRV/SPIRVToLLVMDbgTran.cpp
Comment (jsji) [inline llvm-spirv/lib/SPIRV/SPIRVToLLVMDbgTran.cpp]: This is approved in https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/2628 as well, so we will pulldown it later when it is merged.

File: llvm-spirv/test/extensions/INTEL/SPV_INTEL_runtime_aligned/RuntimeAligned.spt
Comment (AlexeySachkov) [reply llvm-spirv/test/extensions/INTEL/SPV_INTEL_runtime_aligned/RuntimeAligned.spt]: Do you mean that it already fails in SYCL branch? I'm fine with disabling a test temporarily, but we are currently working on making sure that every `XFAIL`ed test has an associated GitHub issue. Submitting one to investigate this failure later and linking it here in the test would be appreciated

File: llvm-spirv/test/extensions/INTEL/SPV_INTEL_cache_controls/multiple-decoration-single-arg.ll
Comment (AlexeySachkov) [inline llvm-spirv/test/extensions/INTEL/SPV_INTEL_cache_controls/multiple-decoration-single-arg.ll]: Tagging @MrSidims here, because this looks like a plain bug in the test, i.e. this change should be upstreamed later.
Both `test` and `test_gep` are kernels and therefore should be represented as `EntryPoint` instruction in SPIR-V. There could be an [`OpName`](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpName) instruction referencing them as well, but it is a part of debug info and it can be safely removed from the module. It is just a coincidence that the test works using it.

I assume that the same change will be applied in the upstream translator and therefore it looks ok to me to proceed with it at intel/llvm

**Review by AlexeySachkov — APPROVED** [review]
Both SPIR-V translator patches LGTM, but as I said in a separate comment, having an issue submitted and linked to an `XFAIL`ed test (2515f6f5e1a3b15c3a9d5ef38421785d6a7541e3) would be very appreciated

File: llvm-spirv/test/extensions/INTEL/SPV_INTEL_runtime_aligned/RuntimeAligned.spt
Comment (bader) [reply llvm-spirv/test/extensions/INTEL/SPV_INTEL_runtime_aligned/RuntimeAligned.spt]: > Do you mean that it already fails in SYCL branch?

The CI checks are to be clean, so I don't think this is the case. @MrSidims added llvm-spirv/test/extensions/INTEL/SPV_INTEL_runtime_aligned/RuntimeAligned.spt test to the Khronos repository a week ago (i.e. this is a new test being pulled to the sycl branch). I think what @jsji means is that some changes in the sycl branch breaks the test. Khronos repository CI validation uses vanilla llvm binaries from llvm/llvm-project repository.

> I'm fine with disabling a test temporarily, but we are currently working on making sure that every `XFAIL`ed test has an associated GitHub issue. Submitting one to investigate this failure later and linking it here in the test would be appreciated

That's right. If we can't fix the problem before the merge someone has to investigate and fix the test after the merge. To make this issue visible, it must be reported.

Comment (jsji) [reply llvm-spirv/test/extensions/INTEL/SPV_INTEL_runtime_aligned/RuntimeAligned.spt]: Sure. https://github.com/intel/llvm/issues/14427 opened to track.

**Comment by jsji** [comment]
> Both SPIR-V translator patches LGTM, but as I said in a separate comment, having an issue submitted and linked to an `XFAIL`ed test ([2515f6f](https://github.com/intel/llvm/commit/2515f6f5e1a3b15c3a9d5ef38421785d6a7541e3)) would be very appreciated

Thanks. https://github.com/intel/llvm/issues/14427 opened for tracking.

File: llvm-spirv/test/extensions/INTEL/SPV_INTEL_runtime_aligned/RuntimeAligned.spt
Comment (jsji) [reply llvm-spirv/test/extensions/INTEL/SPV_INTEL_runtime_aligned/RuntimeAligned.spt]: > I think what @jsji means is that some changes in the sycl branch breaks the test. 

Yes, the newly added spt file asserts in sycl branch.

**Comment by jsji** [comment]
- [ ] Failure in IGC DEV CI Containers / Build and Push IGC Dev Docker Images (Intel Drivers Ubuntu 22.04 Docker image with dev IGC, ubunt... (pull_request) 

This is *known* limitation that we won't be able to download the artifacts if the version is more than 1 weeks' old.  The recent update https://github.com/intel/llvm/pull/14362  was blocked by mismatched igc and compute runtime, hence depending on https://github.com/intel/llvm/pull/14376. @sarnex has been working on https://github.com/intel/llvm/pull/14376.

So this is irrelevant to pulldown, should NOT block the pulldown. @intel/dpcpp-devops-reviewers Can you comments as well? Thanks.

**Review by smanna12 — APPROVED** [review]
CFE changes LGTM

**Comment by jsji** [comment]
@intel/llvm-gatekeepers I think this is ready for merge. 
Last CI run are green:
Pre-commit: https://github.com/intel/llvm/actions/runs/9767449029
Post-commit:  https://github.com/intel/llvm/actions/runs/9767449033

The only change in this round is an include that won't affect any functional tests.

- [ ]  [Remove unecessary include that require c++20](https://github.com/intel/llvm/pull/14327/commits/a15932596c975274ab438f978118f65d7cea2319)

**Comment by elizabethandrews** [comment]
> @intel/dpcpp-cfe-reviewers Would you please have a look at follow testcase updates and template arg fix.
> 
> * [ ]  [[SYCL] Update clang verification checks to use fully qualified class …](https://github.com/intel/llvm/pull/14327/commits/73b3c3f2e4734aa95dbd9d758abb08dde2298b0c)
> * [ ]  [[SYCL] Adjust additional test for clang verification](https://github.com/intel/llvm/pull/14327/commits/f48bae25ce1972790e2781d764a24d826db9cc46)

Looks ok

> * [ ]  [[SYCL][E2E] Fix Wmissing-template-arg-list-after-template-kw warnings](https://github.com/intel/llvm/pull/14327/commits/97e43dddcebfaa44214533888d072fa4cf3f24d5)

Can you share the PR which made this required?

> * [ ]  [[SYCL][Test] Update test due to GEP folding change](https://github.com/intel/llvm/pull/14327/commits/a5f1d633013a35214aad9d47afadbb809ffa523b)

@mdtoguchi can you take a look?

**Comment by jsji** [comment]
> > * [ ]   [[SYCL][E2E] Fix Wmissing-template-arg-list-after-template-kw warnings](https://github.com/intel/llvm/pull/14327/commits/97e43dddcebfaa44214533888d072fa4cf3f24d5)
> 
> Can you share the PR which made this required?

https://github.com/llvm/llvm-project/commit/f46d1463b835560d90ad3ac02b63c771e4ebe566

**Comment by jsji** [comment]
XPASS in SYCL Post Commit / e2e-lin (Intel Arc A-Series Graphics with Level Zero, ["Linux", "arc"], --param matrix-xmx8=True ... / Intel Arc A-Series Graphics with Level Zero (pull_request) is common to others.

**Comment by jsji** [comment]
@intel/llvm-gatekeepers Can someone help to issue a `/merge`. Thanks.

**Comment by steffenlarsen** [comment]
@jsji - There seem to be a conflict that needs to be resolved. The Arc failure should also disappear with a rerun of testing.

**Comment by jsji** [comment]
> @jsji - There seem to be a conflict that needs to be resolved. The Arc failure should also disappear with a rerun of testing.

Arc failure fixed now.  Also pick up fix from @vmaksimo for https://github.com/intel/llvm/issues/14427.

@steffenlarsen @intel/llvm-gatekeepers Can you help to issue a /merge . Thanks.

**Review by bb-sycl — APPROVED** [review]

**Comment by bb-sycl** [comment]
Fri 05 Jul 2024 04:52:20 AM UTC --- Start to merge the commit into sycl branch. It will take several minutes.

**Comment by bb-sycl** [comment]
Fri 05 Jul 2024 04:56:43 AM UTC --- Merge the branch in this PR to base automatically. Will close the PR later.

File: llvm-spirv/test/extensions/INTEL/SPV_INTEL_cache_controls/multiple-decoration-single-arg.ll
Comment (MrSidims) [reply llvm-spirv/test/extensions/INTEL/SPV_INTEL_cache_controls/multiple-decoration-single-arg.ll]: It's not a bug in the test, it's a missing patch in intel/llvm (AFAIK it's being restored, it was missing due to lack of support in IGC).
Basically the missing patch was doing to following: for every kernel function it is creating a wrapper entry point that calls the function with the code copied form the original kernel. It was done because OpenCL allows one kernel to call another kernel and to comply the requirement from SPIR-V spec that one EntryPoint must not call another EntryPoint, we had to create such wrapper (actually, translator's community has done it for us).

**Pull request merged**
